File: solutions/2024/day01/part1.ts
import assert from "assert";

export function solve(input: string): number | string {
  const lines = input.trim().split("\n");

  const leftList: number[] = [];
  const rightList: number[] = [];

  for (const line of lines) {
    const [left, right] = line.split(/\s+/);
    leftList.push(parseInt(left, 10));
    rightList.push(parseInt(right, 10));
  }

  leftList.sort();
  rightList.sort();

  assert.equal(leftList.length, rightList.length);

  let result = 0;

  for (let i = 0; i < leftList.length; i++) {
    const lo = Math.min(leftList[i], rightList[i]);
    const hi = Math.max(leftList[i], rightList[i]);
    const diff = hi - lo;

    result += diff;
  }

  return result;
}

File: solutions/2024/day01/Part1.hs
module Part1 (solve) where

import Data.List (sort)

solve :: String -> String
solve input = show result
  where
    pairs = map (map read . words) $ lines input :: [[Int]]
    (left, right) = let (l, r) = unzip [(a, b) | [a, b] <- pairs] in (sort l, sort r)
    result = sum $ zipWith (\l r -> abs (l - r)) left right
